### PR TITLE
[Flang][Driver][Offload] Support -Xoffload-linker argument in Flang

### PR DESCRIPTION
**Pull request by agozillon**
The -Xoffload-linker command allows forwarding of linker commands to the clang-linker-wrapper used for linking offload libraries into the resulting offload binaries amongst other tasks. This is a rather useful command to have to support the offloading programming models flang-new currently aims to support (OpenMP/OpenACC).

Currently this flag is utilised in the check-offload tests after a recent addition and is used in conjunction with the Fortran OpenMP test suite there, which fails at the moment due to flang-new not recognizing the command, this fixes the issue. The alternative to this would of course be to setup the test config to avoid using this flag with Fortran, but I believe adding support of the flag to flang-new has more merit as having the same compatability/communication capabilities as Clang to the clang-linker-wrapper is important as it's a critical component of the offload pipeline, and the command will likely see more use in the near future.

### CODE REVIEW

**Review by jhuber6 — APPROVED** [review]

**Comment by tblah** [comment]
The CI failures look relevant. The problem isn't immediately apparent to me from the diff (which looks good to me).

**Comment by agozillon** [comment]
hm, weird it passed on my system and I made sure it was running by intentionally failing it, I'll have a look into it and address it! And I'll add the other format as well as an additional test thank you both.

**Comment by agozillon** [comment]
Added the new suggested test and removed the "--" couldn't see anything else that might lead to failure, but we'll see how the CI feels about it, if it doesn't like it please ignore the spam (and my appologies for it) on the PR as I try to work out what the CI wants from me for it to pass as it unfortunately passes locally which makes things hard to track!

File: flang/test/Driver/xoffload-linker.f90 [hunk 1]
```diff
@@ -1,9 +1,10 @@
...
+! RUN: %flang -### --target=x86_64-unknown-linux-gnu -fopenmp --offload-arch=gfx90a -Xoffload-linker a %s 2>&1 | FileCheck %s --check-prefixes=CHECK-XLINKER
 
-! RUN: %flang -### --target=x86_64-unknown-linux-gnu -fopenmp --offload-arch=gfx90a -Xoffload-linker a -Xoffload-linker-amdgcn-amd-amdhsa b %s 2>&1 | FileCheck %s --check-prefixes=CHECK-XLINKER-AMDGCN
+! CHECK-XLINKER: -device-linker=a{{.*}}-
 
-! CHECK-XLINKER-AMDGCN: {{.*}}"--device-linker=a"{{.*}}"--device-linker=amdgcn-amd-amdhsa=b"{{.*}}
+! RUN: %flang -### --target=x86_64-unknown-linux-gnu -fopenmp --offload-arch=gfx90a -Xoffload-linker a -Xoffload-linker-amdgcn-amd-amdhsa b %s 2>&1 | FileCheck %s --check-prefixes=CHECK-XLINKER-AMDGCN
 
-end program
+! CHECK-XLINKER-AMDGCN: -device-linker=a{{.*}}-device-linker=amdgcn-amd-amdhsa=b{{.*}}--
```
Comment (jhuber6) [inline flang/test/Driver/xoffload-linker.f90]: Why is there a `--` here

File: flang/test/Driver/xoffload-linker.f90 [hunk 2]
```diff
@@ -1,9 +1,10 @@
-! RUN %flang -### --target=x86_64-unknown-linux-gnu -fopenmp --offload-arch=gfx90a -Xoffload-linker a %s 2>&1 | FileCheck %s --check-prefixes=CHECK-XLINKER
+! Test the -Xoffload-linker flag that forwards link commands to the clang-linker-wrapper used
+! to help link offloading device libraries
 
-! CHECK-XLINKER {{.*}}--device-linker=a{{.*}}
+! RUN: %flang -### --target=x86_64-unknown-linux-gnu -fopenmp --offload-arch=gfx90a -Xoffload-linker a %s 2>&1 | FileCheck %s --check-prefixes=CHECK-XLINKER
 
-! RUN: %flang -### --target=x86_64-unknown-linux-gnu -fopenmp --offload-arch=gfx90a -Xoffload-linker a -Xoffload-linker-amdgcn-amd-amdhsa b %s 2>&1 | FileCheck %s --check-prefixes=CHECK-XLINKER-AMDGCN
+! CHECK-XLINKER: -device-linker=a{{.*}}-
```
Comment (jhuber6) [inline flang/test/Driver/xoffload-linker.f90]: and a `-` here?

Comment (agozillon) [reply flang/test/Driver/xoffload-linker.f90]: Tried it with multiple variations to see if the CI is happy with it, wouldn't focus too much on what's actually in the test at the moment until I work out the issue the CI has with the test.

Comment (agozillon) [reply flang/test/Driver/xoffload-linker.f90]: e.g. currently. copy pasted a similar passing test at the moment to see if the CI still takes issue with it.

**Comment by agozillon** [comment]
Passed this time, seems the addition of : "{{[^"]*}}clang-linker-wrapper{{.*}}" worked for both the CI and my local machine so that's good. Would love an additional approval from either of you @tblah @banach-space please if possible and the PR seems reasonable of course!

**Review by kiranchandramohan — APPROVED** [review]
LG.

**Pull request merged**